metabase/metabase
# fix funnel tooltip artifacts in safari

Closes #11005

> **I filed a WebKit bug report here:** 🐛
> 
> * bug preview: https://jsbin.com/fapebec/4/edit?html,output
> * bug report: https://bugs.webkit.org/show_bug.cgi?id=257924

### Description

This PR fixes Safari-related rendering artifacts caused by `ChartTooltip` as it moves over `FunnelNormal`.

https://github.com/metabase/metabase/assets/116838/4afd4c77-fc8f-4833-b725-1b9ada39ac91

https://github.com/metabase/metabase/assets/116838/f6363442-0c17-423c-b632-8180492a4dff

Some things I found:

1. Verified that the ChartTooltip was responsible for the artifacts by removing it from the render tree.
2. Adding a background color to `#popover-event-target` fixes the dot trail, but the tooltip hairlines were still showing up (when first creating the query, not after saving).
3. All artifacts disappear when the window size changes, which forces a redraw.
4. Found our internal util function `forceRedraw(element)`.
5. Added a check to only redraw on Desktop Safari.

### How to verify

1.  This depends on [PR #31457](https://github.com/metabase/metabase/pull/31457), so cherry-pick it with this:

    ```
    git cherry-pick -n c91bc16
    ```
3. Open <http://localhost:3000> in Safari
4. New > SQL query > Sample Database
6. Paste query and run:

    ```sql
    SELECT 'Step1' AS Step, 31159 AS Count UNION ALL
    SELECT 'Step2' AS Step, 28372 AS Count UNION ALL
    SELECT 'Step3' AS Step, 21494 AS Count UNION ALL
    SELECT 'Step4' AS Step, 19317 AS Count UNION ALL
    SELECT 'Step5' AS Step, 12284 AS Count UNION ALL
    SELECT 'Step6' AS Step, 1522 AS Count
    ```
7. Visualizations > Funnel
8. Mouse over the Funnel
9. Verify tooltip doesn’t create artifacts.

### Demo

https://github.com/metabase/metabase/assets/116838/3d420950-6186-404a-a759-3b3633a067fb
